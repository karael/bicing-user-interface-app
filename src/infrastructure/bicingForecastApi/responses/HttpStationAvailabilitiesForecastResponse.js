import {
  predictionsType,
  stationIdType,
} from 'infrastructure/bicingForecastApi/responses/httpStationAvailabilitiesForecastResponseType';
import Joi from 'joi';

class HttpStationAvailabilitiesForecastResponse {
  constructor(stationId, availabilities) {
    this.stationId = HttpStationAvailabilitiesForecastResponse.validateStationId(stationId);
    this.availabilities = HttpStationAvailabilitiesForecastResponse.validateAvailabilities(
      availabilities,
    );
  }

  static fromRawValues(stationId, availabilities) {
    return new this(stationId, availabilities);
  }

  static validateStationId(stationId) {
    return Joi.attempt(stationId, stationIdType);
  }

  static validateAvailabilities(availabilities) {
    return Joi.attempt(availabilities, predictionsType);
  }

  // @todo check if add valid model type
  toModelAvailabilities() {
    const model = [];

    this.availabilities.map((availability) => {
      model.push({
        id: this.stationId,
        availableBikeNumber: availability.available_bike_number,
        availableSlotNumber: availability.available_slot_number,
        status: availability.status,
        statedAt: availability.forecast_at,
      });

      return null;
    });

    return model;
  }
}

export default HttpStationAvailabilitiesForecastResponse;
