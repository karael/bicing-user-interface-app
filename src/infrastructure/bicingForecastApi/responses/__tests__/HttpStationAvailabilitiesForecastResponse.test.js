import HttpStationAvailabilitiesForecastResponse
  from 'infrastructure/bicingForecastApi/responses/HttpStationAvailabilitiesForecastResponse';

describe('infrastructure/bicingForecastApi/responses/HttpStationAvailabilitiesForecastResponse', () => {
  test('it can create from raw values', () => {
    const stationId = '76fc88f9-aa4e-4fbf-9852-f23a8cc7c525';
    const availabilities = [
      {
        available_bike_number: 13.5,
        available_slot_number: 12.0,
        forecast_at: '2018-12-24 16:10:00',
        status: 'OPENED',
      },
      {
        available_bike_number: 14.5,
        available_slot_number: 12.667,
        forecast_at: '2018-12-24 16:15:00',
        status: 'OPENED',
      },
    ];

    const response = HttpStationAvailabilitiesForecastResponse.fromRawValues(
      stationId,
      availabilities,
    );

    expect(response.stationId).toEqual(stationId);
    expect(response.availabilities).toEqual(availabilities);
  });
  test('it can not create from raw values with stationId not valid', () => {
    const stationId = 1234;
    const availabilities = [
      {
        available_bike_number: 13.5,
        available_slot_number: 12.0,
        forecast_at: '2018-12-24 16:10:00',
        status: 'OPENED',
      },
    ];

    expect(() => HttpStationAvailabilitiesForecastResponse.fromRawValues(stationId, availabilities))
      .toThrowErrorMatchingSnapshot();
  });
  test('it can not create from raw values with availabilities not valid', () => {
    const stationId = 1234;
    const availabilities = [
      {
        not_valid_key: 14.5,
        available_slot_number: 12.667,
        forecast_at: '2018-12-24 16:10:00',
        status: 'OPENED',
      },
    ];

    expect(() => HttpStationAvailabilitiesForecastResponse.fromRawValues(stationId, availabilities))
      .toThrowErrorMatchingSnapshot();
  });
  test('it can be rendered to availabilities model', () => {
    const stationId = '76fc88f9-aa4e-4fbf-9852-f23a8cc7c525';
    const availabilities = [
      {
        available_bike_number: '13.500000',
        available_slot_number: '12.000000',
        forecast_at: '2018-12-24 16:10:00',
        status: 'OPENED',
      },
      {
        available_bike_number: 14.5,
        available_slot_number: 12.667,
        forecast_at: '2018-12-24 16:15:00',
        status: 'CLOSED',
      },
    ];

    expect(HttpStationAvailabilitiesForecastResponse.fromRawValues(stationId, availabilities)
      .toModelAvailabilities()).toEqual([
      {
        stationId: '76fc88f9-aa4e-4fbf-9852-f23a8cc7c525',
        availableBikeNumber: 13.5,
        availableSlotNumber: 12,
        status: 'OPENED',
        statedAt: '2018-12-24 16:10:00',
      },
      {
        stationId: '76fc88f9-aa4e-4fbf-9852-f23a8cc7c525',
        availableBikeNumber: 14.5,
        availableSlotNumber: 12.667,
        status: 'CLOSED',
        statedAt: '2018-12-24 16:15:00',
      },
    ]);
  });
});
