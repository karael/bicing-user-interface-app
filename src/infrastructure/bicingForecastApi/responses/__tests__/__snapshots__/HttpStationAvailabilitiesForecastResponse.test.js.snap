// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`infrastructure/bicingForecastApi/responses/HttpStationAvailabilitiesForecastResponse it can not create from raw values with availabilities not valid 1`] = `"\\"value\\" must be a string"`;

exports[`infrastructure/bicingForecastApi/responses/HttpStationAvailabilitiesForecastResponse it can not create from raw values with stationId not valid 1`] = `"\\"value\\" must be a string"`;
